pkg/splunk/enterprise/util.go

```diff
@@ -338,117 +343,63 @@ func getAvailableDiskSpace() (uint64, error) {
-		// /opt/splunk/appframework/downloadedApps/default/Standalone/stand1/local/admin/app1.tgz_<hash>
-		localPath := filepath.Join(splcommon.AppDownloadVolume, "downloadedApps", cr.GetNamespace(), kind, cr.GetName(), scope, appSrcName) + "/"
-
-		// create the sub-directories on the volume for downloading scoped apps
-		err = createAppDownloadDir(localPath)
-		if err != nil {
-			return err
-		}
+	vol, err = splclient.GetAppSrcVolume(*appSrc, appFrameworkConfig)
+	if err != nil {
+		scopedLog.Error(err, "unable to get volume spec")
+		return remoteObjectKey, err
+	}
 
-		//TODO: gaurav, handle below two conditions more gracefully in future,
-		// and continue with installation of apps in other app sources.
-		// Get the app source spec to get the volume configuration
-		var vol enterpriseApi.VolumeSpec
-		appSrc, err := getAppSrcSpec(appFrameworkConfig.AppSources, appSrcName)
-		if err != nil {
-			return err
-		}
+	volumePath := vol.Path
+	index := strings.Index(volumePath, "/")
+	volumePath = volumePath[index+1:]
+	location := appSrc.Location
 
-		vol, err = splclient.GetAppSrcVolume(*appSrc, appFrameworkConfig)
-		if err != nil {
-			return err
-		}
+	remoteObjectKey = volumePath + location + appName
 
-		s3ClientWrapper := splclient.S3Clients[vol.Provider]
-		initFunc := s3ClientWrapper.GetS3ClientInitFuncPtr()
-		s3ClientMgr := S3ClientManager{
-			client:          client,
-			cr:              cr,
-			appFrameworkRef: appFrameworkConfig,
-			vol:             &vol,
-			location:        appSrc.Location,
-			initFn:          initFunc,
-			getS3Client:     GetRemoteStorageClient,
-		}
+	return remoteObjectKey, nil
+}
 
-		numOfWorkers := math.Min(float64(appDeployContext.AppsStatusMaxConcurrentAppDownloads), float64(len(appDeployInfo.AppDeploymentInfoList)))
+// getS3ClientMgr gets the S3ClientMgr instance to download apps
+func getS3ClientMgr(client splcommon.ControllerClient, cr splcommon.MetaObject, appFrameworkConfig *enterpriseApi.AppFrameworkSpec, appSrcName string) (*S3ClientManager, error) {
+	scopedLog := log.WithName("getS3ClientMgr").WithValues("crName", cr.GetName(), "namespace", cr.GetNamespace(), "appSrcName", appSrcName)
+	var vol enterpriseApi.VolumeSpec
+	appSrc, err := getAppSrcSpec(appFrameworkConfig.AppSources, appSrcName)
+	if err != nil {
+		scopedLog.Error(err, "unable to get appSrcSpc")
+		return nil, err
+	}
 
-		// handle the app install for each app source.
-		err = handleAppInstallPerAppSrc(client, cr, uint64(numOfWorkers),
-			appSrcName, &appDeployInfo, remoteObjListingMap[appSrcName], s3ClientMgr, localPath, scope)
-		if err != nil {
-			scopedLog.Error(err, "Unable to install apps for app source", "appSource", appSrcName)
-		}
+	vol, err = splclient.GetAppSrcVolume(*appSrc, appFrameworkConfig)
+	if err != nil {
+		scopedLog.Error(err, "unable to get volume spec")
+		return nil, err
 	}
 
-	return err
+	s3ClientWrapper := splclient.S3Clients[vol.Provider]
+	initFunc := s3ClientWrapper.GetS3ClientInitFuncPtr()
```

Comment:
Not a comment on this functionality/MR but the S3 lib in general.  I wonder if we can save the s3Client for future use.  Just thinking out loud but this might be why we see so many IAM connections, because we create a new s3 session every time?

Reply:
I have been thinking about it as well for some time now. Let me create a separate task for it so that we can track it properly.

Reply:
Just a question for  my understanding, is it safe to assume that, once we come out of the scheduler, all the connections are closed?